File: NetDeck/ListCardsViewController.m
//
//  ListCardsViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 10.08.15.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "ListCardsViewController.h"
#import "CardImageViewController.h"
#import "FilterViewController.h"
#import "ImageCache.h"
#import "EditDeckCell.h"

@interface ListCardsViewController ()

@property NSArray* cards;
@property NSArray* sections;

@property CardList* cardList;
@property NSString* filterText;

@property FilterViewController* filterViewController;

@end

static NSString* kSearchFieldValue = @"searchField";
static NSString* kCancelButton = @"cancelButton";

@implementation ListCardsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = l10n(@"Cards");
    
    self.view.backgroundColor = [UIColor colorWithPatternImage:[ImageCache hexTile]];
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.backgroundColor = [UIColor clearColor];
    
    self.statusLabel.font = [UIFont md_systemFontOfSize:13];
    self.statusLabel.text = @"";

    // needed to make the 1 pixel separator show - wtf is this needed here but not elsewhere?
    [self.view bringSubviewToFront:self.toolBar];
    [self.view bringSubviewToFront:self.statusLabel];
    
    [self.tableView registerNib:[UINib nibWithNibName:@"EditDeckCell" bundle:nil] forCellReuseIdentifier:@"cardCell"];
    
    self.cardList = [[CardList alloc] initForRole:self.deck.role];
    
    if (self.deck.identity != nil)
    {
        if (self.deck.role == NRRoleCorp) {
            [self.cardList preFilterForCorp:self.deck.identity];
        } else if (self.deck.role == NRRoleRunner) {
            [self.cardList preFilterForRunner:self.deck.identity];
        }
    }
    
    TableData* data = [self.cardList dataForTableView];
    self.cards = data.values;
    self.sections = data.sections;
    
    NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(showKeyboard:) name:UIKeyboardWillShowNotification object:nil];
    [nc addObserver:self selector:@selector(hideKeyboard:) name:UIKeyboardWillHideNotification object:nil];
    
    UITextField* textField = [self.searchBar valueForKey:kSearchFieldValue];
    if (textField != nil && [textField isKindOfClass:[UITextField class]])
    {
        textField.returnKeyType = UIReturnKeyDone;
    }
    
    [self updateFooter];
}

-(void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
 
    TableData* data = [self.cardList dataForTableView];
    
    self.cards = data.values;
    self.sections = data.sections;
    
    [self.tableView reloadData];
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    NSAssert(self.navigationController.viewControllers.count == 3, @"nav oops");
    
    UINavigationItem* topItem = self.navigationController.navigationBar.topItem;
    
    UIBarButtonItem* filterButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"798-filter-toolbar"] style:UIBarButtonItemStylePlain target:self action:@selector(showFilters:)];
    topItem.rightBarButtonItem = filterButton;
}

-(void) viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    // only clear filter when we're being popped from the nav stack
    if (self.isMovingFromParentViewController) {
        self.searchBar.text = @"";
        self.filterText = nil;
        [self.cardList clearFilters];
    }
}

-(void) dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

-(void) showFilters:(id)sender
{
    if (!self.filterViewController)
    {
        self.filterViewController = [[FilterViewController alloc] initWithNibName:@"FilterViewController" bundle:nil];
    }
    self.filterViewController.role = self.deck.role;
    self.filterViewController.identity = self.deck.identity;
    self.filterViewController.cardList = self.cardList;
    
    // protect against pushing the same controller twice (crashlytics #103)
    if (self.navigationController.topViewController != self.filterViewController)
    {
        [self.navigationController pushViewController:self.filterViewController animated:YES];
    }
}

-(void) countChanged:(UIStepper*)stepper
{
    NSInteger section = stepper.tag / 1000;
    NSInteger row = stepper.tag - (section*1000);
    
    NSArray* arr = self.cards[section];
    Card* card = arr[row];
    
    NSInteger count = 0;
    CardCounter* cc = [self.deck findCard:card];
    if (cc)
    {
        count = cc.count;
    }
    
    NSInteger copies = stepper.value;
    NSInteger diff = ABS(count - copies);
    if (copies < count)
    {
        [self.deck addCard:card copies:-diff];
    }
    else
    {
        [self.deck addCard:card copies:diff];
    }
    
    [self selectTextInSearchBar];
    [self.tableView reloadData];
    
    [self updateFooter];
}

-(void) updateCards
{
    [self.cardList filterByName:self.filterText];
    TableData* data = [self.cardList dataForTableView];
    
    self.cards = data.values;
    self.sections = data.sections;
    [self.tableView reloadData];
    
    [self updateFooter];
}

-(void) updateFooter
{
    NSMutableString* footer = [NSMutableString string];
    [footer appendString:[NSString stringWithFormat:@"%ld %@", (long)self.deck.size, self.deck.size == 1 ? l10n(@"Card") : l10n(@"Cards")]];
    NSString* inf = self.deck.role == NRRoleCorp ? l10n(@"Inf") : l10n(@"Influence");
    if (self.deck.identity && !self.deck.isDraft)
    {
        [footer appendString:[NSString stringWithFormat:@" · %ld/%ld %@", (long)self.deck.influence, (long)self.deck.influenceLimit, inf]];
    }
    else
    {
        [footer appendString:[NSString stringWithFormat:@" · %ld %@", (long)self.deck.influence, inf]];
    }
    
    if (self.deck.role == NRRoleCorp)
    {
        [footer appendString:[NSString stringWithFormat:@" · %ld %@", (long)self.deck.agendaPoints, l10n(@"AP")]];
    }
    
    [footer appendString:@"\n"];
    
    NSArray* reasons = [self.deck checkValidity];
    if (reasons.count > 0)
    {
        [footer appendString:reasons[0]];
    }
    
    self.statusLabel.text = footer;
    self.statusLabel.textColor = reasons.count == 0 ? [UIColor darkGrayColor] : [UIColor redColor];
}

#pragma mark search bar

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    self.filterText = searchText;
    [self updateCards];
}

-(void) searchBarTextDidEndEditing:(UISearchBar *)searchBar
{
    searchBar.showsCancelButton = NO;
    self.tableView.tableHeaderView = self.searchBar;
}

-(void) searchBarTextDidBeginEditing:(UISearchBar *)searchBar
{
    searchBar.showsCancelButton = YES;
    UIButton* cancelBtn = [self.searchBar valueForKey:kCancelButton];
    if (cancelBtn) {
        [cancelBtn setTitle:l10n(@"Done") forState:UIControlStateNormal];
    }
    self.tableView.tableHeaderView = self.searchBar;
}

-(void) searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [searchBar resignFirstResponder];
}

-(void) searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    [self selectTextInSearchBar];
    
    if (self.cards.count > 0)
    {
        Card* card = [self.cards objectAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]];
        if (card)
        {
            [self.deck addCard:card copies:1];
            [self.tableView reloadData];
            
            [self updateFooter];
        }
    }
}

-(void) selectTextInSearchBar
{
    UITextField* textField = [self.searchBar valueForKey:kSearchFieldValue];
    if (textField == nil || ![textField isKindOfClass:[UITextField class]])
    {
        return;
    }
    
    [textField setSelectedTextRange:[textField textRangeFromPosition:textField.beginningOfDocument toPosition:textField.endOfDocument]];
}

#pragma mark - tableview

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSArray* arr = self.cards[section];
    return arr.count;
}

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.sections.count;
}

-(NSString*)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSArray* arr = self.cards[section];
    return [NSString stringWithFormat:@"%@ (%ld)", self.sections[section], (long)arr.count];
}

-(void) tableView:(UITableView *)tableView willDisplayCell:(EditDeckCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    cell.backgroundColor = [UIColor whiteColor];
    
    Card* card = [self.cards objectAtIndexPath:indexPath];
    CardCounter* cc = [self.deck findCard:card];
    
    if (cc.count > 0)
    {
        cell.nameLabel.font = [UIFont md_mediumSystemFontOfSize:16];
    }
    else
    {
        cell.nameLabel.font = [UIFont md_systemFontOfSize:16];
    }
}

-(UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    EditDeckCell* cell = [tableView dequeueReusableCellWithIdentifier:@"cardCell" forIndexPath:indexPath];
    
    cell.stepper.tag = indexPath.section * 1000 + indexPath.row;
    [cell.stepper addTarget:self action:@selector(countChanged:) forControlEvents:UIControlEventValueChanged];
    
    Card* card = [self.cards objectAtIndexPath:indexPath];
    CardCounter* cc = [self.deck findCard:card];
    
    cell.stepper.minimumValue = 0;
    cell.stepper.maximumValue = card.maxPerDeck;
    cell.stepper.value = cc.count;
    cell.stepper.hidden = NO;
    cell.idButton.hidden = YES;
    
    if (cc)
    {
        if (card.unique)
        {
            cell.nameLabel.text = [NSString stringWithFormat:@"%lu× %@ •", (unsigned long)cc.count, card.name];
        }
        else
        {
            cell.nameLabel.text = [NSString stringWithFormat:@"%lu× %@", (unsigned long)cc.count, card.name];
        }
    }
    else
    {
        if (card.unique)
        {
            cell.nameLabel.text = [NSString stringWithFormat:@"%@ •", card.name];
        }
        else
        {
            cell.nameLabel.text = [NSString stringWithFormat:@"%@", card.name];
        }
    }
    
    
    NSString* type = [Faction name:card.faction];;
    
    NSInteger influence = [self.deck influenceFor:cc];
    if (cc.count == 0 && self.deck.identity.faction != card.faction)
    {
        influence = card.influence;
    }
    
    if (influence > 0)
    {
        cell.influenceLabel.text = [NSString stringWithFormat:@"%ld", (long)influence];
        cell.influenceLabel.textColor = card.factionColor;
    }
    else
    {
        cell.influenceLabel.text = @"";
    }
    
    if (card.isMostWanted) {
        cell.mwlLabel.text = [NSString stringWithFormat:@"%ld", (long)MIN(-1, -cc.count)];
    }
    cell.mwlLabel.hidden = !card.isMostWanted;
    
    NSString* subtype = card.subtype;
    if (subtype)
    {
        // type = [type stringByAppendingString:influenceStr];
        type = [type stringByAppendingString:@" · "];
        type = [type stringByAppendingString:card.subtype];
        cell.typeLabel.text = type;
    }
    else
    {
        // type = [type stringByAppendingString:influenceStr];
        cell.typeLabel.text = type;
    }

    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Card* card = [self.cards objectAtIndexPath:indexPath];
    
    CardImageViewController* img = [[CardImageViewController alloc] initWithNibName:@"CardImageViewController" bundle:nil];
    img.cards = [self.cardList allCards];
    img.selectedCard = card;
    
    [self.navigationController pushViewController:img animated:YES];
}

#pragma mark - keyboard

-(void) showKeyboard:(NSNotification*)notification
{
    CGRect kbRect = [[notification.userInfo objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
    float kbHeight = kbRect.size.height;
    
    UIEdgeInsets inset = self.tableView.contentInset;
    inset.bottom = kbHeight;
    self.tableView.contentInset = inset;
    self.tableView.scrollIndicatorInsets = inset;
}

-(void) hideKeyboard:(id)notification
{
    UIEdgeInsets inset = self.tableView.contentInset;
    inset.bottom = 0;
    self.tableView.contentInset = inset;
    self.tableView.scrollIndicatorInsets = inset;
}

@end

File: NetDeck/ListCardsViewController.swift
//
//  ListCardsViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 17.12.16.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import UIKit
import SwiftyUserDefaults

final class ListCardsViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, UISearchBarDelegate {
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var searchBar: UISearchBar!
    
    @IBOutlet weak var summaryLabel: UILabel!
    @IBOutlet weak var statusLabel: TickingLabel!
    
    @IBOutlet weak var toolBar: UIToolbar!
    
    private var keyboardObserver: KeyboardObserver!
    
    var deck: Deck!
    
    private var cards = [[Card]]()
    private var sections = [String]()
    private var cardList: CardList!
    private var filterText = ""
    private var searchScope = CardSearchScope.name
    
    private var filterViewController: FilterViewController!
    
    private let kSearchFieldValue = "searchField"
    private let kCancelButton = "cancelButton"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Cards".localized()
        
        self.view.backgroundColor = UIColor(patternImage: ImageCache.hexTile)
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        self.tableView.backgroundColor = .clear
        
        self.statusLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 13, weight: UIFont.Weight.regular)
        self.statusLabel.text = ""
        self.summaryLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 13, weight: UIFont.Weight.regular)
        self.summaryLabel.text = ""
        
        // needed to make the 1 pixel separator show - wtf is this needed here but not elsewhere?
        self.view.bringSubviewToFront(self.toolBar)
        self.view.bringSubviewToFront(self.statusLabel)
        self.view.bringSubviewToFront(self.summaryLabel)
        
        self.tableView.register(UINib(nibName: "EditDeckCell", bundle: nil), forCellReuseIdentifier: "cardCell")
        
        let packUsage = Defaults[.deckbuilderPacks]
        self.cardList = CardList(role: self.deck.role, packUsage: packUsage, browser: false, legality: self.deck.legality)
        
        if let identity = self.deck.identity {
            if self.deck.role == .corp {
                self.cardList.preFilterForCorp(identity, self.deck.mwl)
            } else if self.deck.role == .runner {
                self.cardList.preFilterForRunner(identity, self.deck.mwl)
            }
        }
        
        self.keyboardObserver = KeyboardObserver(handler: self)
        
        if let textField = self.searchBar.value(forKey: kSearchFieldValue) as? UITextField {
            textField.returnKeyType = .done
        }
        
        self.tableView.scrollFix()

        if self.traitCollection.forceTouchCapability == .available {
            self.registerForPreviewing(with: self, sourceView: self.view)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        let data = self.cardList.dataForTableView()
        self.sections = data.sections
        self.cards = data.values
        
        self.tableView.reloadData()
        self.updateFooter()
        
        let filterButton = UIBarButtonItem(image: UIImage(named: "798-filter-toolbar"), style: .plain, target: self, action: #selector(self.showFilters(_:)))
        let scopeButton = UIBarButtonItem(image: UIImage(named: "708-search-toolbar"), style: .plain, target: self, action: #selector(self.changeScope(_:)))
        
        self.navigationItem.rightBarButtonItems = [filterButton, scopeButton]
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        self.tableView.flashScrollIndicators()
    }
        
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        if self.isMovingFromParent {
            self.searchBar.text = ""
            self.filterText = ""
            self.cardList.clearFilters()
        }
    }
    
    @objc func showFilters(_ sender: UIBarButtonItem) {
        if self.filterViewController == nil {
            self.filterViewController = FilterViewController()
        }
        self.filterViewController.role = self.deck.role
        self.filterViewController.identity = self.deck.identity
        self.filterViewController.cardList = self.cardList
        
        if self.navigationController?.topViewController != self.filterViewController {
            self.navigationController?.pushViewController(self.filterViewController, animated: true)
        }
    }
    
    @objc func countChanged(_ stepper: UIStepper) {
        let section = stepper.tag / 1000
        let row = stepper.tag - (section * 1000)
        
        let card = self.cards[section][row]
        let cc = self.deck.findCard(card)
        let count = cc?.count ?? 0
        
        let copies = Int(stepper.value)
        let diff = copies - count
        self.deck.addCard(card, copies: diff)
        
        self.selectTextInSearchBar()
        self.tableView.reloadData()
        self.updateFooter()
    }
    
    func updateCards() {
        switch self.searchScope {
        case .name: self.cardList.filterByName(self.filterText)
        case .text: self.cardList.filterByText(self.filterText)
        case .all: self.cardList.filterByTextOrName(self.filterText)
        }
        
        let data = self.cardList.dataForTableView()
        self.sections = data.sections
        self.cards = data.values
        
        self.tableView.reloadData()
        self.updateFooter()
    }
    
    func updateFooter() {
        var summary = ""
        summary = String(format: "%ld %@", self.deck.size, self.deck.size == 1 ? "Card".localized() : "Cards".localized())
        let inf = self.deck.role == .corp ? "Inf".localized() : "Influence".localized()
        if self.deck.identity != nil && !self.deck.isDraft {
            summary += String(format: " · %ld/%ld %@", self.deck.influence, self.deck.influenceLimit, inf)
        } else {
            summary += String(format: " · %ld %@", self.deck.influence, inf)
        }
        
        if self.deck.role == .corp {
            summary += String(format: " · %ld %@", self.deck.agendaPoints, "AP".localized())
        }
        
        let reasons = self.deck.checkValidity()
        let status = reasons.first ?? "Deck is valid".localized()
        
        self.summaryLabel.text = summary
        self.summaryLabel.textColor = reasons.count == 0 ? .systemGray : .systemRed
            
        self.statusLabel.text = status
        self.statusLabel.strings = reasons
        self.statusLabel.textColor = reasons.count == 0 ? .systemGray : .systemRed
    }
    
    // MARK: - search bar
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        self.filterText = searchText
        self.updateCards()
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        searchBar.setShowsCancelButton(false, animated: true)
    }
    
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        searchBar.setShowsCancelButton(true, animated: true)
        if let button = searchBar.value(forKey: kCancelButton) as? UIButton {
            button.setTitle("Done".localized(), for: .normal)
        }
    }
    
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        self.selectTextInSearchBar()
        
        if self.cards.count > 0 {
            if let card = self.cards.first?.first {
                self.deck.addCard(card, copies: 1)
                self.tableView.reloadData()
                
                self.updateFooter()
            }
        }
    }
    
    func selectTextInSearchBar() {
        if let textField = self.searchBar.value(forKey: kSearchFieldValue) as? UITextField {
            let range = textField.textRange(from: textField.beginningOfDocument, to: textField.endOfDocument)
            textField.selectedTextRange = range
        }
    }
    
    @objc func changeScope(_ btn: UIBarButtonItem) {
        let alert = UIAlertController(title: "Search in:".localized(), message: nil, preferredStyle: .alert)
        
        alert.addAction(UIAlertAction(title: "Name".localized().checked(self.searchScope == .name)) { action in
            self.searchScope = .name
            self.updateCards()
        })
        alert.addAction(UIAlertAction(title: "Text".localized().checked(self.searchScope == .text)) { action in
            self.searchScope = .text
            self.updateCards()
        })
        alert.addAction(UIAlertAction(title: "All".localized().checked(self.searchScope == .all)) { action in
            self.searchScope = .all
            self.updateCards()
        })
        alert.addAction(UIAlertAction.alertCancel(nil))
        
        self.present(alert, animated: true, completion: nil)
    }
    
    // MARK: - tableview
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.cards[section].count
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.sections.count
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return self.sections[section]
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cardCell", for: indexPath) as! EditDeckCell
        
        cell.stepper.tag = indexPath.section * 1000 + indexPath.row
        cell.stepper.addTarget(self, action: #selector(self.countChanged(_:)), for: .valueChanged)
        
        let card = self.cards[indexPath.section][indexPath.row]
        let cardCounter = self.deck.findCard(card)
        
        cell.stepper.minimumValue = 0
        cell.stepper.maximumValue = Double(card.maxPerDeck)
        cell.stepper.value = Double(cardCounter?.count ?? 0)
        cell.stepper.isHidden = false
        cell.idButton.isHidden = true
        
        let weight: CGFloat
        let text: String
        if let cc = cardCounter {
            text = cc.displayName(self.deck.mwl)
            weight = UIFont.Weight.medium.rawValue
        } else {
            text = card.displayName(self.deck.mwl)
            weight = UIFont.Weight.regular.rawValue
        }
        cell.nameLabel.text = text
        cell.nameLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 16, weight: UIFont.Weight(rawValue: weight))
        cell.nameLabel.textColor = card.owned == 0 ? .systemGray : .label

        let faction = Faction.name(for: card.faction)
        let subtype = card.subtype
        if subtype.count > 0 {
            let type = faction + " · " + subtype
            cell.typeLabel.text = type
        } else {
            cell.typeLabel.text = faction
        }
        
        let cc = cardCounter ?? CardCounter(card: card, count: 1)
        let influence = self.deck.influenceFor(cc)
        
        if influence > 0 {
            cell.influenceLabel.text = "\(influence)"
            cell.influenceLabel.textColor = card.factionColor
        } else {
            cell.influenceLabel.text = ""
        }
        
        cell.mwlLabel.text = ""
        let penalty = card.mwlPenalty(self.deck.mwl)
        let list = MWLManager.mwlBy(self.deck.mwl)
        if penalty > 0 && !list.universalInfluence {
            cell.mwlLabel.text = "\(min(-1, -cc.count * penalty))"
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let card = self.cards[indexPath.section][indexPath.row]
        
        let img = CardImageViewController()
        img.setCards(self.cardList.allCards(), mwl: self.deck.mwl, deck: self.deck)
        img.selectedCard = card
        
        self.navigationController?.pushViewController(img, animated: true)
    }
}

extension ListCardsViewController: UIViewControllerPreviewingDelegate {
    func previewingContext(_ previewingContext: UIViewControllerPreviewing, commit viewControllerToCommit: UIViewController) {
        if let imgController = viewControllerToCommit as? CardImageViewController {
            imgController.peeking = false
        }
        self.show(viewControllerToCommit, sender: self)
    }

    func previewingContext(_ previewingContext: UIViewControllerPreviewing, viewControllerForLocation location: CGPoint) -> UIViewController? {
        let cellPosition = self.tableView.convert(location, from: self.view)
        guard
            let indexPath = self.tableView.indexPathForRow(at: cellPosition),
            let cell = tableView.cellForRow(at: indexPath)
        else {
            return nil
        }

        let card = self.cards[indexPath.section][indexPath.row]

        let imgController = CardImageViewController(peeking: true)
        imgController.setCards(self.cardList.allCards(), mwl: self.deck.mwl, deck: self.deck)
        imgController.selectedCard = card
        imgController.preferredContentSize = CGSize(width: 0, height: 436)

        previewingContext.sourceRect = self.tableView.convert(cell.frame, to: self.view)

        return imgController
    }
}

extension ListCardsViewController: KeyboardHandling {

    // MARK: - keyboard
    
    func keyboardWillShow(_ info: KeyboardInfo) {
        let screenHeight = UIScreen.main.bounds.size.height
        let kbHeight = screenHeight - info.endFrame.origin.y - self.toolBar.frame.size.height
        
        var inset = self.tableView.contentInset
        inset.bottom = kbHeight
        self.tableView.contentInset = inset
        self.tableView.scrollIndicatorInsets = inset
    }
    
    func keyboardWillHide(_ info: KeyboardInfo) {
        var inset = self.tableView.contentInset
        inset.bottom = 0
        self.tableView.contentInset = inset
        self.tableView.scrollIndicatorInsets = inset
    }
    
}
